### PR TITLE
PR: [WIP] Add maximize pane button inside dockwidget title bar

### LINKED ISSUE #1165: Console tabs are too big because they display the full filename
_From [andreasn...@googlemail.com](https://code.google.com/u/106330885117898830116/) on 2012-10-04T15:20:39Z_

The console tabs of running or run programs are quite big, because they display the complete filename including path, e.g.:

"Console - /home/user/python/myproject/util/module.py"

On some Spyder window setups, this big tab uses much space and thus leaves very little space for the other views tabs, I have put in the same pane. These other tabs then often show only one letter and it is somehow unpleasing to the eye.

Not so long ago, only the filename ("Console - module.py") was displayed in these tabs. Can we have an option for choosing between long and short filenames in the console tabs?

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1165_

### LINKED ISSUE #2272: Add a maximize button for each pane
Currently if we have to maximize a pane then we have to click on the button on the main toolbar
screenshot from 2015-03-25 20 00 43

A new button can be added right next to the two buttons in the screenshot above, and we can provide help text when a user hovers over the buttons

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #2272 
Fixes #1165

# Todo:
- [x] Test vertical title bar
- [x] Make floating panes independent windows
- [x] Make title visible when maximized
- [x] Add option to remove panes title bar (while docked and with locked option on)
- [x] Add docstrings
- [x] Update icons

### CODE REVIEW

**Comment by goanpeca** [comment]
Fixes #2272

Todo:
- [x] Test vertical title bar
- [x] Update icons
- [ ] Remove maximize button from toolbar?
- [ ] Make title visible when maximized?
- [ ] Add docstring

**Comment by goanpeca** [comment]
This is still not ready for merging but I would like to see if I can get some feedback in the meantime.

Can you test this @thewhitetulip

**Comment by goanpeca** [comment]
@ccordoba12 I am having a small issue with the CustomTitleBarWidget when the Style is changed in the preferences... 

The look:
**Horizontal**
image

**Vertical**
image

**Comment by ccordoba12** [comment]
Mmmh, I don't like too much. There is way too much space around widgets.

Can something be done about it?

**Comment by goanpeca** [comment]
@ccordoba12 as I said, this is not ready for merging... of course the idea is to make it look the best, but that is the least difficult part of this PR, I will improve on the look later on

(take a look at the code first, ;-)  )

**Comment by goanpeca** [comment]
And to answer your question, yes we can do anything we want, now is a custom title bar :),

**Comment by ccordoba12** [comment]
Then let's see how it evolves. The code seems a bit too much for just adding one button, but I we can make more things with it (like centering the plugin tittle), I'd agree with it.

I remember that sometime ago you mentioned that you had several improvements for how our dock widgets look. Do you plan to submit a PR for that too? I think that's more important (at least to start with) than this button.

**Comment by goanpeca** [comment]
Yes, that is why I said that Qt should provide a simpler way of doing this, but unfortunately it does not, that is why there is so much boiler plate code to get the custom title bar widget going. 

On the up side, yes, we can now do many more things (like what you suggest). this PR tackles partially a part of improving on the look of the gui, so this is a start, another PR will come later to do the rest.

**Comment by thewhitetulip** [comment]
255 lines of code needs to be added for one button?

**Comment by goanpeca** [comment]
@thewhitetulip yes, cause the title bar in a dockwidget is a fixed thing, so any type of customization on it requires defining a complete new custom widget.

**Comment by thewhitetulip** [comment]
Oh, so in this case, if we need to add any more items then the code need not be this huge correct?

**Comment by goanpeca** [comment]
Yes.. from now on we can customize the title bar of a dock widget with few additional code. But this huge amount was needed as boilerplate...

**Comment by thewhitetulip** [comment]
Okay, that is nice :+1:

**Comment by Nodd** [comment]
Can't you just subclass qt's titlebar widget to add the button ? It would require less code (but maybe it's not possible).

**Comment by goanpeca** [comment]
@nodd nope it is not possible

File: spyderlib/plugins/__init__.py [hunk 1]
```diff
@@ -120,9 +335,17 @@ class SpyderDockWidget(QDockWidget):
     plugin_closed = Signal()
 
     def __init__(self, *args, **kwargs):
+        # This is needed for the new maximize button in the custom title bar
+        self.get_focus_wiget = kwargs.pop('get_focus_func')
```
Comment (Nodd) [inline spyderlib/plugins/__init__.py]: idget with a 'd' here

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Thanks hawkeye :)

**Comment by goanpeca** [comment]
Not sure if it is worth all the trouble to have this at this moment. Even if the maximize button per panel makes sense, the way spyder behaves right now is:

The widget of a maximized dockwidget is placed as the central widget, by doing this, the central widget looses the title bar (cause it is no longer a dockwidget... but only the widget inside), so there is no unmaximize button, cause... there is no title bar.

The only way I can see to make this work, would be to make a base class to inherit from when making the main widgets. This way the base class could add a title bar to the layout when the layout is set (by overloading...).

Unless we are ok on not having the title of a maximized panel and not having the buttons... which would require to press on the button in the toolbar to go back... but that would defeat the original purpose.

Comments?

**Comment by Nodd** [comment]
Or change the maximize behavior by hiding all other dockwidgets ?
BTW while testing the current behavior I found out thaat the outline explorer was not hidden if another widget was maximized.

**Comment by goanpeca** [comment]
Or do that... which would seem like less trouble :smile:

**Comment by goanpeca** [comment]
This is deeper than originally envisioned. Closing

**Comment by thewhitetulip** [comment]
@goanpeca I might sound naive, is this PR merged or was it closed without merge?

**Comment by goanpeca** [comment]
@thewhitetulip it was closed without merge since it raised more problems than it would solve, at this point I do not see any purpose in continuing with this.

**Comment by ccordoba12** [comment]
@goanpeca, I'd still like to see your (promised) improvements to how our dock widgets look. As I said before, I think that's more important than this work :-)

**Comment by goanpeca** [comment]
Sure @ccordoba12, but that is for another PR ;-)

**Comment by ccordoba12** [comment]
Sure!! ;-)

**Comment by Nodd** [comment]
Changing the way dockwidgets look would mean that Spyder won't follow the style defined by the user. Do we want that ?

**Comment by goanpeca** [comment]
@Nodd it would follow the style, but we could tweak the design as to make it look the best withing the spyder context, so it should not be a problem

**Comment by goanpeca** [comment]
Due to discussions in #2353, this PR seems relevant enough to introduce the base code for some of the mentioned fixes.

**Comment by goanpeca** [comment]
To be able to continue with this I would like some feedback @ccordoba12.

## Maximize

When maximizing a dockwidget, right now the widget is move as the central widget of the main application, and hence the title bar of the dockwidget is not present anymore. If the aim is to add a maximize button inside the dockwidget titlebar, it seems weird that this title bar disappears and the user has to click on a different icon to restore the view os Spyder.

1.) Would it make sense to include the TitleBar (the same as the dockwidget) when the widget is maximized, so that the user can restore clicking on the same button he/she pressed to maximize?

## Floating 
I have been trying some options to get a floating widget to behave as an independent window so that it can be resized and maximized usng meta keys and arrows in windows.

**Linux**
On linux an additional entry in the taskbar is created, but double clicking on the window title does not restore the dockwidget back to spyder application, so we would need to have a button for this... which feels weird.....

**Windows**
On windows the only way to have an additional tarkbar entry is to assign the dockwidget to a new QMainWindow. But double clicking on the title bar restores the dockwidget back to spyder application, which is quite nice and intuitive.

2.) Is it worth it to create a new QMainWindow when floating a dockwidget to have an additional  taskbar on windows?

**Comment by ccordoba12** [comment]
My answers:

> Would it make sense to include the TitleBar (the same as the dockwidget) when the widget is maximized?

Yes, it seems quite reasonable.

> On linux an additional entry in the taskbar is created, but double clicking on the window title does not restore the dockwidget back to spyder application.

I'd prefer a button, so that the UI be clear and easy to grasp for our users. It's hard to predict how many of them are going to discover that a double click on the windows' title will restore the widget to the main window.

> Is it worth it to create a new QMainWindow when floating a dockwidget to have an additional taskbar on windows?

If an additional entry is shown on Linux, we should show it on Windows too (for consistency :-)

**Comment by goanpeca** [comment]
@ccordoba12 thanks for the answers...

So, I think I have a working PR now and I will post a video soon to show how it operates, but I still need to fix the default maximize in spyder 'cause it hides any floating widget as well, which defeats the purpose of having a floating widget.

Regarding the issue of having a separate taskbar entry in windows, it would make the code much harder to implement to take care of the second window reparenting plus some methods (I have seen this in the object inspector) rely on querying the parent with parent(), which breaks when a dockwidget was reparented. So, in windows the default window manager behavior is to NOT make an extra entry.. that is how windows works, in linux is different that is how it works.

**Comment by goanpeca** [comment]
Here is a [video](http://youtu.be/hK6CSUbArWg) of the new maximize button in operation + floating dockwidgets as windows so they can be arranged using meta keys and arrows (in windows)

Notice that the dockwidget buttons are disabled as needed and also when maximizing a widget now it has a titlebar as well.

**Comment by blink1073** [comment]
:clap:  very cool @goanpeca.

**Comment by thewhitetulip** [comment]
Awesome :+1:

**Comment by goanpeca** [comment]
I added a new option to hide panes title bars when locked

image

And this is how it looks (@d1manson)

image

Now the question, @ccordoba12, @Nodd, would it make sense to have the pane options (float, maximize, close) in 1.) the context menu of the tab, or 2.) in the pane menu 
(image).

For 1 the problem is that if the panes is alone, there is no tab to attach a context menu to....

For 2, the pane menu is something that we need to make sure every pane has and in the same position, not the case right now (but that will be another PR where I will try to homogenize the look of spyder panes.

**Comment by Nodd** [comment]
If the titlebars are hidden then it makes sense to not provide options to close or float them (closing can still be done from the pane menu). For maximize I would just add a shortcut to maximize the current widget.
So at the end I think there is no need to add these options anywhere!

**Comment by goanpeca** [comment]
There is already a `maximize current pane` in view. So in that line of thinking either we remove it... now a pane can be maximized on its own, or we add a new entry to `float current pane`? I would prefer to remove them.

**Comment by Nodd** [comment]
Why do you want a `float current pane` action so much for a *locked* pane ?
Maximize can be kept because maximizing the editor while coding then unmaximizing for testing can be useful even with locked panes.

**Comment by goanpeca** [comment]
The idea of 'locked pane' was to avoid accidental clicking and dragging not to remove the options from the pane

**Comment by goanpeca** [comment]
I might want to have the title bars hidden to maximize the space, and still I might want to float a given pane as an independent window... and I should be able to do it, is not that I want it so much, is that that functionality should not be lost

**Comment by Nodd** [comment]
Well, unlock, move your pane, relock ?

Anyway, some power user will one day open an issue to add a keybind for floating od docking a pane, we might as well add it now...

**Comment by d1manson** [comment]
@goanpeca - screenshots/videos are looking good to me.  Thanks for doing all these #2353-related PRs!

**Comment by goanpeca** [comment]
Ok rebasing this is getting toooo tricky so I think I will close this one and start a new PR!

**Comment by blink1073** [comment]
wimp!  :stuck_out_tongue_winking_eye:

**Comment by goanpeca** [comment]
Maybe :smile:

**Comment by Nodd** [comment]
What I often do is rebase commit by commit. Most commit won't conflicts so it will be very fast, and when you have a conflict only the changes for the next commit are shown, instead of all changes since your fork. So it's a bit slow but you're sure to not forget anything.
I don't know if git has such a feature build in (like "rebase until next conflict").

**Comment by goanpeca** [comment]
@blink1073 got it :)

**Comment by goanpeca** [comment]
@SylvainCorlay I need some help with the following icons, do we have something already in place for normal spyder icons and QtAwesome icons?

- Maximize
- Float
- Close

screen

**Comment by ccordoba12** [comment]
Any progress on this one? @goanpeca, is this something you plan to finish?

**Comment by goanpeca** [comment]
Yes, I need to rebase and should be almost ready for merging.

**Comment by goanpeca** [comment]
@ccordoba12 after a big squash and a long rebase this one is ready for more reviewing.

@Nodd could you start your nitpicking :-) ?

**Pull request closed without merging**